modules/_labs/pagination/react/lib/useIsMobile.tsx

```diff
@@ -0,0 +1,24 @@
+import React, {useEffect, useState} from 'react';
+
+const useIsMobile = (container?: React.RefObject<HTMLElement>) => {
+  const [mobile, setMobile] = useState(false);
+
+  const widthIsMobile = () =>
+    container && container.current && container.current.clientWidth
+      ? container.current.clientWidth
+      : window.innerWidth;
+
+  useEffect(() => {
```

Comment:
Did you experience a rapid flash using `useEffect` vs `useLayoutEffect`? `useEffect` will happen after a frame has been flushed to the DOM and could cause flashing and more layout thrashing when mobile is true (quite possibly worst-case for lower-power mobile devices.

`useEffect` is a good default, especially given that using `useLayoutEffect` to change DOM can lead to cascading read/write DOM cycles which makes frames more expensive.

This whole hook looks like it might be a problem... First, `mobile` defaults to `false`. When the container ref changes (usually always after a render has happened), the real calculation happens after another frame has rendered. At that point a resize handler is used against the window object, but the calculation is against the container clientWidth...

So the implementation is mixed. It will default to the `window.innerWidth` if the ref hasn't been set by React yet. If it has, it will use the `clientWidth` of the container. After that, it no longer cares about the container width anymore and will re-evaluate when the `window` size is changed. So is `isMobile` determined by the width of the browser or the width of the container? Should we just pick one? @lychyi

Reply:
I went ahead and cleaned up this hook to only look at the window inner width. The customization by passing your own container was confusing and the logic wasn't completely correct.

Reply:
This is better. As it is right now, `setState` is called on _every_ resize, not just when `mobile` switches from `false` to `true` or vice versa. Better performance would be to call `setMobile` only when `mobile` changes for its previous value.

I do wonder though if we could actually handle window resizing without using en event listener. If the `Pages` component renders both wide and compressed DOM and then uses a media query to set `display: none` on the non-matching size. The trade-off is not having an additional event listener. but gaining extra VDOM comparisons.

I'm not sure we decided how to handle parent container sizes. So far we're avoiding them. There are many libraries (both React-based and non-react) and most of them don't scale well and get stuck in DOM read/write layout thrashing.

I remember doing a custom solution based on [fastdom](https://github.com/wilsonpage/fastdom) to avoid multiple forced reflows. No matter what, it is hard to scale non-native element resize events and should only be used when absolutely necessary and that should probably be controlled by the application.

I think flexbox and css grid are the best way we can make components responsive and resize tricks should be avoided.

Reply:
It seems hooks have a reference-based check when notifying render functions, so calling `setMobile` seems to have no performance issue